kotlinx-coroutines-core/common/src/channels/BufferedChannel.kt

```diff
@@ -0,0 +1,3087 @@
+                    // This `receive()` operation should suspend.
+                    if (waiter === null) {
+                        // The waiter is not specified;
+                        // return the corresponding result.
+                        return SUSPEND_NO_WAITER
+                    }
+                    // Try to install the waiter.
+                    if (segment.casState(index, state, waiter)) {
+                        // The waiter has been successfully installed.
+                        // Invoke the `expandBuffer()` procedure and finish.
+                        expandBuffer()
+                        return SUSPEND
+                    }
+                }
+            }
+            // The cell stores a buffered element.
+            state === BUFFERED -> if (segment.casState(index, state, DONE_RCV)) {
+                // Retrieve the element and expand the buffer.
+                expandBuffer()
+                return segment.retrieveElement(index)
+            }
+        }
+        return updateCellReceiveSlow(segment, index, r, waiter)
+    }
+
+    private fun updateCellReceiveSlow(
+        /* The working cell is specified by
+        the segment and the index in it. */
+        segment: ChannelSegment<E>,
+        index: Int,
+        /* The global index of the cell. */
+        r: Long,
+        /* The waiter to be stored in case of suspension. */
+        waiter: Any?,
+    ): Any? {
+        // The cell state should be updated according to  its state machine;
+        // see the paper mentioned in the very beginning for the algorithm details.
+        while (true) {
+            // Read the current cell state.
+            val state = segment.getState(index)
+            when {
+                // The cell is empty.
+                state === null || state === IN_BUFFER -> {
+                    // If a rendezvous must happen, the operation does not wait
+                    // until the cell stores a buffered element or a suspended
+                    // sender, poisoning the cell and restarting instead.
+                    // Otherwise, try to store the specified waiter in the cell.
+                    val senders = sendersAndCloseStatus.value.sendersCounter
+                    if (r < senders) {
+                        // The cell is already covered by sender,
+                        // so a rendezvous must happen. Unfortunately,
+                        // the cell is empty, so the operation poisons it.
+                        if (segment.casState(index, state, POISONED)) {
+                            // When the cell becomes poisoned, it is essentially
+                            // the same as storing an already cancelled receiver.
+                            // Thus, the `expandBuffer()` procedure should be invoked.
+                            expandBuffer()
+                            return FAILED
+                        }
+                    } else {
```

Comment:
That's because you were lucky enough (make `isStressTest = true` and run the Lincheck tests, they will likely fail). 

First, the intuition is that the `update...Slow(..)` methods cover all transitions in the state machine. Second, in this particular place, the cell could be updated from `null` to `IN_BUFFER`, so the transition fails in the fast path, and the operation goes to the same branch in the slow path.